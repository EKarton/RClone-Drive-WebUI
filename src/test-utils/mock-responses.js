export const mockFiles = {
  list: [
    {
      Path: '.rclone',
      Name: '.rclone',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-11-01T21:09:05.237Z',
      IsDir: true,
      ID: '1a4mI4vsGKypqFFPwLnrFid1ZdvlhirDi',
    },
    {
      Path: 'Documents',
      Name: 'Documents',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T01:51:05.982Z',
      IsDir: true,
      ID: '1_za5Kqy9ZjXZ7rsztGpKhtJLTuNCKv8q',
    },
    {
      Path: 'Games',
      Name: 'Games',
      Size: -9,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T05:14:59.237Z',
      IsDir: true,
      ID: '1j6mAjxQoJokiU8yWJYgyWYIiSIgH1G7M',
    },
    {
      Path: 'Literature',
      Name: 'Literature',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-29T07:10:05.059Z',
      IsDir: true,
      ID: '1dj1T5YkjRQZNiQ5efCA8PzideJb88d3K',
    },
    {
      Path: 'Pictures',
      Name: 'Pictures',
      Size: -10,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:39:11.850Z',
      IsDir: true,
      ID: '1D2PpmoIT23NqyTnolmohZ0UxvOd2D7n4',
    },
    {
      Path: 'Software',
      Name: 'Software',
      Size: -2,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T03:45:02.471Z',
      IsDir: true,
      ID: '1zzr0Bwp9tel66rCUKz8mZIXRBR2kuywr',
    },
    {
      Path: 'Videos',
      Name: 'Videos',
      Size: -8,
      MimeType: 'inode/directory',
      ModTime: '2021-09-28T04:22:24.997Z',
      IsDir: true,
      ID: '1eJlJ1PT7b9XY3Q5tLeU49BNkug259986',
    },
    {
      Path: 'backup.sh',
      Name: 'backup.sh',
      Size: 1110,
      MimeType: 'application/x-sh',
      ModTime: '2021-11-14T18:58:44.655Z',
      IsDir: false,
    },
  ],
};

export const mockRemotes = {
  remotes: ['googledrive', 'onedrive', 'icloud'],
};

export const mockPictures = {
  list: [
    {
      Path: 'Pictures/2012/20120517_171428.JPG',
      Name: '20120517_171428.JPG',
      Size: 335408,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T21:14:30.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_135206.JPG',
      Name: '20120517_135206.JPG',
      Size: 262719,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T17:52:06.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_204557.JPG',
      Name: '20120516_204557.JPG',
      Size: 245109,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T00:45:58.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_174510.JPG',
      Name: '20120517_174510.JPG',
      Size: 271914,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T21:45:12.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120820_144833.JPG',
      Name: '20120820_144833.JPG',
      Size: 2662880,
      MimeType: 'image/jpeg',
      ModTime: '2012-08-20T18:48:32.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120517_094409.JPG',
      Name: '20120517_094409.JPG',
      Size: 280596,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-17T13:44:10.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_181954.JPG',
      Name: '20120516_181954.JPG',
      Size: 416424,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-16T22:19:54.000Z',
      IsDir: false,
    },
    {
      Path: 'Pictures/2012/20120516_181809.JPG',
      Name: '20120516_181809.JPG',
      Size: 211575,
      MimeType: 'image/jpeg',
      ModTime: '2012-05-16T22:18:10.000Z',
      IsDir: false,
    },
  ],
};
