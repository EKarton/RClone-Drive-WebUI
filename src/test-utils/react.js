import { FileViewerProvider } from 'contexts/FileViewerStore';
import { RCloneInfoStateProvider } from 'contexts/RCloneInfoStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState, initialRCloneInfoState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  const renderedComponent = render(
    <RCloneInfoStateProvider defaultState={initialRCloneInfoState}>
      <FileViewerProvider defaultState={initialFileViewerState}>
        <Router history={history}>{component}</Router>
      </FileViewerProvider>
    </RCloneInfoStateProvider>
  );

  return {
    ...renderedComponent,
    history,
  };
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
