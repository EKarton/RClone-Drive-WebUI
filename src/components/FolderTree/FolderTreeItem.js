import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import TreeItem from '@mui/lab/TreeItem';
import PropTypes from 'prop-types';
import { useEffect, useState } from 'react';
import useRCloneClient from 'hooks/rclone/useRCloneClient';
import { StatusTypes } from 'utils/constants';

export default function FolderTreeItem({ remote, curPath, label }) {
  const rCloneClient = useRCloneClient();

  const [status, setStatus] = useState(null);
  const [subFolders, setSubFolders] = useState();

  const [isExpanded, setIsExpanded] = useState(false);

  useEffect(() => {
    const fetchFolders = async () => {
      try {
        setStatus(StatusTypes.LOADING);

        const data = await rCloneClient.fetchSubFolders(remote, curPath);

        setSubFolders(data.sort());
        setStatus(StatusTypes.SUCCESS);
      } catch (err) {
        setStatus(StatusTypes.ERROR);
      }
    };

    if (isExpanded) {
      fetchFolders();
    }
  }, [curPath, isExpanded, rCloneClient, remote]);

  const handleTreeItemClicked = () => {
    setIsExpanded(true);
  };

  const renderSubFolders = () => {
    if (!isExpanded || !status) {
      return <div />;
    }

    if (status === StatusTypes.LOADING) {
      return <div>Loading</div>;
    }

    if (status === StatusTypes.ERROR) {
      return <div>Error!</div>;
    }

    if (subFolders.length === 0) {
      return <div>No subfolders</div>;
    }

    return subFolders.map((subFolder) => (
      <FolderTreeItem
        remote={remote}
        curPath={subFolder.Path}
        label={subFolder.Name}
        key={subFolder.Name}
      />
    ));
  };

  return (
    <TreeItem
      nodeId={`${remote}:${curPath}`}
      label={label}
      onClick={handleTreeItemClicked}
      expandIcon={<ChevronRightIcon />}
      collapseIcon={<ExpandMoreIcon />}
    >
      {renderSubFolders()}
    </TreeItem>
  );
}

FolderTreeItem.propTypes = {
  remote: PropTypes.string,
  curPath: PropTypes.string,
  label: PropTypes.string,
};
