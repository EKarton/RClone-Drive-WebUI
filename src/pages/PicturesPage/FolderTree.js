import TreeView from "@mui/lab/TreeView";
import TreeItem from "@mui/lab/TreeItem";
import ExpandMoreIcon from "@mui/icons-material/ExpandMore";
import ChevronRightIcon from "@mui/icons-material/ChevronRight";
import useRCloneClient from "hooks/useRCloneClient";
import { useEffect, useState } from "react";
import { StatusTypes } from "pages/PicturesListPage";

const FolderTree = ({ remotes, onFolderSelect }) => {
  const handleNodeSelect = (_e, nodeIds) => {
    onFolderSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} />
      ))}
    </TreeView>
  );
};

const FolderTreeItem = ({ remote, curPath, label }) => {
  const rCloneClient = useRCloneClient();

  const [status, setStatus] = useState(StatusTypes.LOADING);
  const [subFolders, setSubFolders] = useState(undefined);
  const [error, setError] = useState(null);

  const [isExpanded, setIsExpanded] = useState(false);

  useEffect(() => {
    const fetchFolders = async () => {
      try {
        setError(null);
        setStatus(StatusTypes.LOADING);

        const data = await rCloneClient.fetchSubFolders(remote, curPath);

        setSubFolders(data.sort());
        setStatus(StatusTypes.SUCCESS);
      } catch (err) {
        setError(err);
      }
    };

    if (isExpanded) {
      fetchFolders();
    }
  }, [curPath, isExpanded, rCloneClient, remote]);

  const handleTreeItemClicked = () => {
    setIsExpanded(true);
  };

  const renderSubFolders = () => {
    if (!isExpanded) {
      return <div />;
    }

    if (status === StatusTypes.LOADING) {
      return <div>Loading</div>;
    }

    if (status === StatusTypes.ERROR) {
      return <div>Error! {error}</div>;
    }

    if (subFolders.length === 0) {
      return <div>No subfolders</div>;
    }

    return subFolders.map((subFolder) => (
      <FolderTreeItem
        remote={remote}
        curPath={subFolder.Path}
        label={subFolder.Name}
      />
    ));
  };

  return (
    <TreeItem
      nodeId={`${remote}:${curPath}`}
      label={label}
      onClick={handleTreeItemClicked}
      expandIcon={<ChevronRightIcon />}
      collapseIcon={<ExpandMoreIcon />}
    >
      {renderSubFolders()}
    </TreeItem>
  );
};

export default FolderTree;
