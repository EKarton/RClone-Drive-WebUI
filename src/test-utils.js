import { FileViewerProvider, initialState } from 'contexts/FileViewerStore';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { createMemoryHistory } from 'history';
import { Router } from 'react-router';

const customRender = (
  component,
  { initialFileViewerState = initialState } = {},
  { route = '/' } = {}
) => {
  const history = createMemoryHistory({ initialEntries: [route] });

  return render(
    <FileViewerProvider defaultState={initialFileViewerState}>
      <Router history={history}>{component}</Router>
    </FileViewerProvider>
  );
};

export * from '@testing-library/react';
export * from '@testing-library/user-event';
export { customRender, userEvent };
