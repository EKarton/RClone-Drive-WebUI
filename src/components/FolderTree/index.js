import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import TreeView from '@mui/lab/TreeView';
import PropTypes from 'prop-types';
import FolderTreeItem from './FolderTreeItem';

export default function FolderTree({ remotes, onSelect }) {
  const handleNodeSelect = (_e, nodeIds) => {
    onSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} key={remote} />
      ))}
    </TreeView>
  );
}

FolderTreeItem.propTypes = {
  remotes: PropTypes.arrayOf(PropTypes.string),
  onSelect: PropTypes.func,
};
