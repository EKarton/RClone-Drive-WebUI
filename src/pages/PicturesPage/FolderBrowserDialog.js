import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';
import { useState } from 'react';
import FolderTree from 'components/FolderTree';

/**
 * Represents the dialog that appears when a user clicks on a remote
 */
export default function FolderBrowserDialog({ remotes, open, onCancel, onOk, title }) {
  const [remotePath, setRemotePath] = useState('');

  const handleFolderTreeSelected = (remotePath) => {
    setRemotePath(remotePath);
  };

  const handleOk = () => {
    if (!remotePath) {
      onCancel();
      return;
    }

    onOk(remotePath);
  };

  return (
    <Dialog
      sx={{ '& .MuiDialog-paper': { width: '80%', maxHeight: 435 } }}
      maxWidth="xs"
      open={open}
      onClose={onCancel}
    >
      <DialogTitle>{title}</DialogTitle>
      <DialogContent dividers>
        <FolderTree remotes={remotes} onSelect={handleFolderTreeSelected} />
      </DialogContent>
      <DialogActions>
        <Button autoFocus onClick={onCancel} data-testid="cancel-button">
          Cancel
        </Button>
        <Button onClick={handleOk} data-testid="ok-button">
          Ok
        </Button>
      </DialogActions>
    </Dialog>
  );
}
