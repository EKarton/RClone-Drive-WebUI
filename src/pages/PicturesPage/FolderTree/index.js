import TreeView from '@mui/lab/TreeView';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import FolderTreeItem from './FolderTreeItem';

export default function FolderTree({ remotes, onFolderSelect }) {
  const handleNodeSelect = (_e, nodeIds) => {
    onFolderSelect(nodeIds);
  };

  return (
    <TreeView
      aria-label="file tree navigator"
      defaultCollapseIcon={<ExpandMoreIcon />}
      defaultExpandIcon={<ChevronRightIcon />}
      onNodeSelect={handleNodeSelect}
    >
      {remotes.map((remote) => (
        <FolderTreeItem remote={remote} curPath="" label={remote} />
      ))}
    </TreeView>
  );
}
