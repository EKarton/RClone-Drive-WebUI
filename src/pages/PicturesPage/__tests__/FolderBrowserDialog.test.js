import FolderTree from 'components/FolderTree';
import { mockRemotes } from 'test-utils/mock-responses';
import { act, customRender, userEvent, screen } from 'test-utils/react';
import FolderBrowserDialog from '../FolderBrowserDialog';

jest.mock('components/FolderTree');

describe('FolderBrowserDialog', () => {
  it('should render FolderTree and call handleOk() once correctly if user selects an item from the FolderTree', async () => {
    // Mock user selecting a folder
    jest.useFakeTimers();

    FolderTree.mockImplementation(({ onSelect }) => {
      setTimeout(() => onSelect('googledrive:Pictures'), 10000);
      return null;
    });

    const onOkFn = jest.fn();
    const onCancelFn = jest.fn();
    const { baseElement } = customRender(
      <FolderBrowserDialog
        remotes={mockRemotes.remotes}
        open
        onOk={onOkFn}
        onCancel={onCancelFn}
        title="Title"
      />
    );

    expect(baseElement).toMatchSnapshot();

    act(() => jest.advanceTimersByTime(10000));

    userEvent.click(screen.getByTestId('ok-button'));

    expect(onOkFn).toBeCalledTimes(1);
    expect(onOkFn).toBeCalledWith('googledrive:Pictures');
    expect(onCancelFn).not.toBeCalled();
  });

  it('should call handleCancel() when the Folder Tree clicks on onSelect() with no data', async () => {
    // Mock user selecting a folder
    jest.useFakeTimers();

    FolderTree.mockImplementation(({ onSelect }) => {
      setTimeout(() => onSelect(''), 10000);
      return null;
    });

    const onOkFn = jest.fn();
    const onCancelFn = jest.fn();
    customRender(
      <FolderBrowserDialog
        remotes={mockRemotes.remotes}
        open
        onOk={onOkFn}
        onCancel={onCancelFn}
        title="Title"
      />
    );

    act(() => jest.advanceTimersByTime(10000));

    userEvent.click(screen.getByTestId('ok-button'));

    expect(onCancelFn).toBeCalledTimes(1);
    expect(onOkFn).not.toBeCalled();
  });

  it('should call handleCancel() when the user clicks on the Cancel button', async () => {
    FolderTree.mockReturnValue(null);

    const onOkFn = jest.fn();
    const onCancelFn = jest.fn();
    customRender(
      <FolderBrowserDialog
        remotes={mockRemotes.remotes}
        open
        onOk={onOkFn}
        onCancel={onCancelFn}
        title="Title"
      />
    );

    userEvent.click(screen.getByTestId('cancel-button'));

    expect(onCancelFn).toBeCalledTimes(1);
    expect(onOkFn).not.toBeCalled();
  });
});
