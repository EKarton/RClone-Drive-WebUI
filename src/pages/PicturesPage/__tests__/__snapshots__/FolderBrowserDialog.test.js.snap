// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FolderBrowserDialog should render FolderTree and call handleOk() once correctly if user selects an item from the FolderTree 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiModal-root MuiDialog-root css-1lxu39g-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthXs css-22jxwj-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
        >
          Title
        </h2>
        <div
          class="MuiDialogContent-root MuiDialogContent-dividers css-1t4vnk2-MuiDialogContent-root"
        />
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-hlj6pa-MuiDialogActions-root"
        >
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root Mui-focusVisible  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="cancel-button"
            tabindex="0"
            type="button"
          >
            Cancel
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            >
              <span
                class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
          <button
            class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            data-testid="ok-button"
            tabindex="0"
            type="button"
          >
            Ok
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
